ratis-server/src/main/java/org/apache/ratis/server/impl/ReadRequests.java

```diff
@@ -92,9 +99,15 @@ Consumer<Long> getAppliedIndexConsumer() {
   }
 
   CompletableFuture<Long> waitToAdvance(long readIndex) {
-    if (stateMachine.getLastAppliedTermIndex().getIndex() >= readIndex) {
-      return CompletableFuture.completedFuture(readIndex);
+    final long lastApplied = stateMachine.getLastAppliedTermIndex().getIndex();
+    if (lastApplied >= readIndex) {
+      return CompletableFuture.completedFuture(lastApplied);
+    }
+    final CompletableFuture<Long> f = readIndexQueue.add(readIndex);
+    final long current = stateMachine.getLastAppliedTermIndex().getIndex();
+    if (current >= lastApplied) {
```

Comment:
This should be `current > lastApplied`; otherwise, the condition is always true.

The change looks good other than that.

Reply:
You are right! I was thinking about `current >= readIndex`. Thanks very much for the careful reviews ;)